### PR TITLE
catalog: Add more structure to state init writes

### PR DESCRIPTION
**Pull request by jkosh44**
Previously, initializing the in-memory catalog state involved a mixture
of reading and writing from and to the durable catalog state. This
commit adds some more structure to the writes involved in this process.
The writes are not removed completely, but they are grouped into three
distinct groups instead of mixed throughout the process. Those groups
are:

  1. Before any data is read into the in-memory catalog state, new
     builtin objects are written down durably and removed builtin
     objects are removed from the durable state. Currently, we don't
     check all types of builtin objects for removal, and it's likely
     that certain types are leaked in the durable state. This stage
     can/should arguably be done at a lower level and should be
     completely handled by the durable catalog without any involvement
     from the in-memory catalog. That refactor may be done in a later
     commit. Durably recording system variables from LaunchDarkly
     piggybacks off this stage and is included out of convenience.
  2. After all data is read into the in-memory catalog state, user item
     ASTs are migrated, if any migrations are defined.
  3. After user item ASTs are migrated, all builtin items whose schema
     has changed since the last version are migrated.

This structure subjectively makes the state initialization easier to
reason about and more importantly will make parts of the
multi-subscriber catalog easier to write.

Works towards resolving #24844

### Motivation
This PR refactors existing code.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

**Comment by jkosh44** [comment]
@ParkMyCar adding you as a reviewer in case you have bandwidth today since @mjibson is out. If not, no worries just ignore this and I'll just wait until tomorrow.

File: src/adapter/src/catalog/open.rs [hunk 1]
```diff
@@ -296,47 +301,27 @@ impl Catalog {
...
-            // storage "config" collection so that we can toggle the flag with
-            // Launch Darkly, but use it in boot before Launch Darkly is available.
-            txn.set_persist_txn_tables(state.system_config().persist_txn_tables())?;
-            Catalog::load_builtin_types(&mut state, &mut txn)?;
-
-            let persisted_builtin_ids: BTreeMap<_, _> = txn
-                .get_system_object_mappings()
-                .map(|mapping| (mapping.description, mapping.unique_identifier))
+            // Load all builtin types.
+            let (builtin_types, builtin_non_types): (Vec<_>, Vec<_>) = builtin_item_ids.all_builtins.into_iter().partition(|(builtin, _)| matches!(builtin, Builtin::Type(_)));
```
Comment (ParkMyCar) [inline src/adapter/src/catalog/open.rs:309]: do you mind breaking these onto new lines? unless this is what `rustfmt` automatically does
```suggestion
            let (builtin_types, builtin_non_types): (Vec<_>, Vec<_>) = builtin_item_ids
                .all_builtins
                .into_iter()
                .partition(|(builtin, _)| matches!(builtin, Builtin::Type(_)));
```

**Review by ParkMyCar — APPROVED** [review]
LGTM

File: src/adapter/src/catalog/open.rs (hunk 1 above)
Comment (jkosh44) [reply src/adapter/src/catalog/open.rs]: Done, `rustfmt` seems to be completely broken on this file for some reason. I've noticed a ton of unformatted code that is not fixed by `rustfmt`.

**Pull request merged**